### PR TITLE
Updates to auto-updater on Windows

### PR DESCRIPTION
**Pull request by CookiePLMonster**
This PR brings several improvements to Updater, focusing mostly on Windows specific improvements (but not only those). Here's the list:

- **Generic**: Fixed an issue where `Updater::OnUpdateAvailable` invoked `m_parent->close()` from a wrong thread - debug builds threw an assertion error there.
- **Generic**: Removed manual `UI::Stop()` calls in favour of a scope guard. While this does not seem to matter much for Mac, changes to Windows specific logic made it extremely important that this function is called in all cases - so made this change to remove any risk that updater terminates without having called this function.
- **Windows**: Modified the flow of updater's UI thread and simplified synchronization between said thread and main thread. Now UI thread no longer spinlocks, instead relies on `GetMessage` to sleep when there is no new messages. Additionally, synchronization has been remade to account for this fact (no need to poll `request_stop`), with UI thread being properly joinable now - this has removed the need for a second synchronization flag. \
Thanks to this change, on my PC (with 8 logical cores) CPU usage of `Updater.exe` when idle has dropped from around 13-14% to nearly 0%.
- **Windows**: Spawning `Updater.exe` process kept process/thread handles opened in `Dolphin.exe` - while this doesn't matter much since this process is terminating shortly after spawning the updater, it's a good practice to close those handles when not using them. Also in a very unlikely case of `Updater.exe` terminating and `Dolphin.exe` still running (I don't know if this is actually possible) updater will no longer linger as a zombie process.
- **Windows**: Fixed a `STARTUPINFO sinfo = {sizeof(info)};` typo which could have had unknown consequences - worst case trashing the stack if `CreateProcess` behaviour changes in the future!
- **Windows**: Removed a `const_cast` in a call to `CreateProcess` - as of C++17, a non-const overload to `data()` exists which fits this use case perfectly.
- **Windows**: Replaced a `GetModuleFileName` call in WinUpdater with a custom `GetModuleName` function which doesn't impose a `MAX_PATH` limit. Also, the old call had a wrong size parameter passed, as size parameter in `GetModuleFileName` is supposed to be in characters, not in bytes.

### CODE REVIEW

**Comment by leoetlino** [comment]
Your PR description should be part of the commit message. The current one is way too vague and it would also be nice if you could split that commit into several smaller commits to make reviewing and bisecting easier.

**Comment by CookiePLMonster** [comment]
All done! Thankfully I had those separated and only squashed all commits for submission - so I only reworded commit messages to be better.

Each change is also separate and works well standalone, so the separation is good.

**Comment by CookiePLMonster** [comment]
Consider this PR on hold, I found one more possible improvement for Windows updater code (removing arbitrary path length limit if my initial assessment is right) and I'd like to include it here.

**Comment by stenzek** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/WinUpdater/WinUI.cpp [hunk 1]
```diff
@@ -199,49 +200,40 @@ void SetDescription(const std::string& text)
...
-  std::thread thread(MessageLoop);
-  thread.detach();
+  ui_thread = std::thread(MessageLoop);
 }
 
 void Stop()
 {
-  if (!running.IsSet())
-    return;
+  if (window_handle != nullptr)
```
Comment (stenzek) [inline Source/Core/WinUpdater/WinUI.cpp]: `if (window_handle)`

File: Source/Core/WinUpdater/WinUI.cpp [hunk 2]
```diff
@@ -259,8 +251,11 @@ void Sleep(int sleep)
 void WaitForPID(u32 pid)
 {
   HANDLE parent_handle = OpenProcess(SYNCHRONIZE, FALSE, static_cast<DWORD>(pid));
-  WaitForSingleObject(parent_handle, INFINITE);
-  CloseHandle(parent_handle);
+  if (parent_handle != nullptr)
```
Comment (stenzek) [inline Source/Core/WinUpdater/WinUI.cpp]: `if (parent_handle)`

File: Source/Core/WinUpdater/Main.cpp [hunk 3]
```diff
@@ -32,6 +32,28 @@ std::vector<std::string> CommandLineToUtf8Argv(PCWSTR command_line)
   LocalFree(tokenized);
   return argv;
 }
+
+std::optional<std::wstring> GetModuleName(HINSTANCE hInstance)
```
[the remark this answers is not in the document]
Comment (CookiePLMonster) [reply Source/Core/WinUpdater/Main.cpp]: Won't hurt to do so, true.

**Comment by altimumdelta** [comment]
Updater.exe is a bit of a (too) vague name, no? Maybe do something about that while you're at it.

**Comment by JosJuice** [comment]
I don't think that's something we need to change in this PR.

**Comment by altimumdelta** [comment]
Allright, sure no need, maybe in future.

**Comment by CookiePLMonster** [comment]
I don't necessarily agree, why is it vague?

**Comment by altimumdelta** [comment]
I meant from an OS view, anything could be an "updater" on a system, makes the taskmanager less of a pain if names are more descriptive of it's origin/parent/author.

**Comment by CookiePLMonster** [comment]
But that's not the problem of an executable name - instead, metadata should be added if it's not added yet - so in Processes tab it'd show as "Dolphin Updater" or so.

That's of course assuming it's not done this way already, I can't test right now.

**Comment by stenzek** [comment]
> I meant from an OS view, anything could be an "updater" on a system, makes the taskmanager less of a pain if names are more descriptive of it's origin/parent/author.

That's a fair point. But you can also find out which updater it is by opening the file properties in task manager, or in the processes tab, it'd show the full description.

**Comment by altimumdelta** [comment]
> But that's not the problem of an executable name - instead, metadata should be added if it's not added yet - so in Processes tab it'd show as "Dolphin Updater" or so.

Ah, people like me, that deal with processes usually use the raw Details view or another utility which mostly does things with raw process names, not the new Win10 process labels, but that should help some people I guess.
